src/QsCompiler/TextProcessor/CompletionParsing.fs

```diff
@@ -0,0 +1,230 @@
+﻿module Microsoft.Quantum.QsCompiler.TextProcessing.CompletionParsing
+
+open System
+open FParsec
+open Microsoft.Quantum.QsCompiler.TextProcessing.Keywords
+open Microsoft.Quantum.QsCompiler.TextProcessing.ParsingPrimitives
+open Microsoft.Quantum.QsCompiler.TextProcessing.SyntaxBuilder
+open Microsoft.Quantum.QsCompiler.TextProcessing.TypeParsing
+
+
+/// Describes the environment of a code fragment in terms of what kind of completions are available.
+type CompletionEnvironment =
+    /// The code fragment is inside a namespace but outside any callable.
+    | NamespaceTopLevel
+
+/// Describes the kind of identifier that is expected at a particular position in the source code.
+type IdentifierKind =
+    /// The identifier is a new symbol declaration.
+    | Declaration
+    /// The identifier is a type name.
+    | Type
+    /// The identifier is a characteristic set name.
+    | Characteristic
+    /// The identifier is the given keyword.
+    | Keyword of string
+    /// The identifier is a namespace.
+    | Namespace
+    /// The identifier is a member of the given namespace and has the given kind.
+    | Member of string * IdentifierKind
+
+/// Parses the end-of-transmission character.
+let private eot =
+    pchar '\u0004'
+
+/// Parses a symbol. Includes reserved keywords if the keyword is followed immediately by EOT without any whitespace
+/// (i.e., a possibly incomplete symbol that happens to start with a keyword). Does not consume whitespace.
+let private symbol =
+    let qsIdentifier =
+        identifier <| IdentifierOptions (isAsciiIdStart = (fun c -> isAsciiLetter c || c = '_'),
+                                            isAsciiIdContinue = isSymbolContinuation)
+    notFollowedBy qsReservedKeyword >>. qsIdentifier <|>
+    (qsIdentifier .>> previousCharSatisfiesNot Char.IsWhiteSpace .>> followedBy eot)
```

Comment:
the previousCharSatisfiesNot seems redundant if you are not consuming any whitespace in any case?